dotnet/src/Connectors/Connectors.Memory.CosmosNoSql/CosmosNoSqlServiceCollectionExtensions.cs

```diff
@@ -1,198 +1,335 @@
+    /// <param name="options">Optional options to further configure the <see cref="CosmosNoSqlVectorStore"/>.</param>
+    /// <param name="lifetime">The service lifetime for the store. Defaults to <see cref="ServiceLifetime.Singleton"/>.</param>
+    /// <returns>Service collection.</returns>
+    [RequiresUnreferencedCode(DynamicCodeMessage)]
+    [RequiresDynamicCode(UnreferencedCodeMessage)]
+    public static IServiceCollection AddKeyedCosmosNoSqlVectorStore(
+        this IServiceCollection services,
+        object? serviceKey,
+        CosmosNoSqlVectorStoreOptions? options = default,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
     {
-        // If we are not constructing Database, add the IVectorStore as transient, since we
-        // cannot make assumptions about how Database is being managed.
-        services.AddKeyedTransient<VectorStore>(
-            serviceId,
-            (sp, obj) =>
-            {
-                var database = sp.GetRequiredService<Database>();
-                options ??= sp.GetService<CosmosNoSqlVectorStoreOptions>() ?? new()
-                {
-                    EmbeddingGenerator = sp.GetService<IEmbeddingGenerator>()
-                };
-
-                return new CosmosNoSqlVectorStore(database, options);
-            });
+        Verify.NotNull(services);
+
+        services.Add(new ServiceDescriptor(typeof(CosmosNoSqlVectorStore), serviceKey, (sp, _) =>
+        {
+            var database = sp.GetRequiredService<Database>();
+            options = GetStoreOptions(sp, _ => options);
+
+            return new CosmosNoSqlVectorStore(database, options);
+        }, lifetime));
+
+        services.Add(new ServiceDescriptor(typeof(VectorStore), serviceKey,
+            static (sp, key) => sp.GetRequiredKeyedService<CosmosNoSqlVectorStore>(key), lifetime));
 
         return services;
     }
 
     /// <summary>
-    /// Register an Azure CosmosDB NoSQL <see cref="VectorStore"/> with the specified service ID
-    /// and where the Azure CosmosDB NoSQL <see cref="Database"/> is constructed using the provided <paramref name="connectionString"/> and <paramref name="databaseName"/>.
+    /// Registers a <see cref="CosmosNoSqlVectorStore"/> as <see cref="VectorStore"/>
+    /// using the provided <paramref name="connectionString"/> and <paramref name="databaseName"/>.
     /// </summary>
-    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="VectorStore"/> on.</param>
+    /// <inheritdoc cref="AddKeyedCosmosNoSqlVectorStore(IServiceCollection, object?, string, string, CosmosNoSqlVectorStoreOptions?, ServiceLifetime)"/>
+    [RequiresUnreferencedCode(DynamicCodeMessage)]
+    [RequiresDynamicCode(UnreferencedCodeMessage)]
+    public static IServiceCollection AddCosmosNoSqlVectorStore(
+        this IServiceCollection services,
+        string connectionString,
```

Comment:
We should probably check (later) if this makes sense for Cosmos - IIRC a connection string isn't the regular/recommended way to connect (there's an endpoint and an auth token or something).

I think we should generally keep the CosmosNoSql provider prerelease (there are other issues), so we can fix this later.